api/models/error.go

```diff
@@ -161,6 +161,11 @@ var (
 		error: errors.New("Unable to service the request for the reservation period"),
 	}
 
+	ErrRunnerUnavailable = err{
+		code:  http.StatusBadGateway,
+		error: errors.New("Received a gRPC error Unavailable from a runner."),
+	}
```

Comment:
should we include some idea of Retry-After headers for errors like these, where the intention is to retry on the client? the list here is mostly intended to be user-facing errors, now there's lb-facing and user-facing errors, these seem like two separate things (for example, this error is not intended to show to the user and is very much more of an 'internal' server error for us, that shouldn't get surfaced but we should log). without retry headers (afaik the preferred way to signal to client to retry) it's unclear that this error should be retried on the client side without parsing the string message out, and it seems easier to add retry-after for any relevent error rather than build up a set of error messages to retry. also seems like in that case this should be 503 instead of 502, RFC: https://tools.ietf.org/html/rfc7231#section-6.6.4

so i'm kinda thinking for errors like these we want to tack on the ability to add a retry header without displaying the message in the response. if this is completely off base, then just let me know. here's what I'm thinking:

```go
type RetryableError interface {
  error
  RetryAfter() int // seconds 
}

type retryableError struct {
  error
  delay int
}

// no need to override Error() we don't want to change the error on these
func (r *retryableError) RetryAfter() int { return r.delay }

func NewRetryableError(e error) RetryableError {
  return &retryableError{e, 1*time.Second} // some default, pick one...
}

func NewRetryableErrorDelay(e error, delay int) RetryableError {
  return &retryableError{e, delay}
}
```

and https://github.com/fnproject/fn/blob/master/api/server/error_response.go would after checking for models.APIError (which the idea here with the retryable error is that it would _not_ be one of these, ie we just hand off the original error to NewRetryableError and log it just the same as an internal server error), it could check for retryable, eg:

```go
switch x := err.(type); x {
case models.APIError:
  // same
case models.RetryableError:
  status = 503
  w.Header().Set("Retry-After", strconv.Itoa(x.RetryAfter())
default:
}

// TODO log it and then write response (log retryable same as internal server error)
```

we're already doing the retry-after thing https://github.com/fnproject/fn/blob/master/api/server/error_response.go#L48 but we can clean it up and stop using the api errors as a stand in for them, I think. what does anyone make of all that? I'd be willing to write this up if things are unclear to clarify them, it should not be too much code. the main thing is not using api errors for retryable internal errors at the lb level, so that the api errors only serve one purpose...

thanks for proposing this, just gave me the above idea after looking around. let me know what y'all think